README.org
* Try Using Emacs and Org-mode
To learen use emacs fllowing [[https://cestlaz.github.io/stories/emacs/][Mike Zamansky]]


eshell/history
pwd
ls
find-file Git-and-Github.org 
ls
tree
mv learning-how-to-learn.org Computer-Science/
tree
exit
ping www.baidu.com
ping mirrors.tuna.tsinghua.edu.cn
ls
chmod +x pand.sh 
ls
cd notes/
ls
cd Computer-Science/
ls
./pand.sh 
emacs pand.sh 
./pand.sh 


samples/python/test.py
def puls(s):
    return puls+1



init.el
;;add elpa.org
(require 'package)
(setq package-enable-at-startup nil)
(add-to-list 'package-archives
	     '("melpa" . "http://elpa.emacs-china.org/melpa/")
	     '("gnu"   . "http://elpa.emacs-china.org/gnu/"))

;; (setq package-archives '(("gnu"   . "http://elpa.emacs-china.org/gnu/")
;;  			 ("melpa" . "http://elpa.emacs-china.org/melpa/")))

(package-initialize)

;; Bootstrap `use-package'
(unless (package-installed-p 'use-package)
  (package-refresh-contents)
  (package-install 'use-package))



(org-babel-load-file (expand-file-name "~/.emacs.d/myinit.org"))
(custom-set-variables
 ;; custom-set-variables was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(org-agenda-files (quote ("~/Dropbox/org/inbox.org")))
 '(org-confirm-babel-evaluate nil)
 '(org-default-notes-file (concat org-directory "/notes.org"))
 '(org-directory "~/Dropbox/org")
 '(org-export-html-postamble nil)
 '(org-hide-leading-stars t)
 '(org-image-actual-width nill)
 '(org-src-fontify-natively t)
 '(org-startup-folded (quote overview))
 '(org-startup-indented t)
 '(package-selected-packages
   (quote
    (org-pdftools pdf-tools pdf-tool org-pdfview org-pomodoro emmet-mode counsel-projectile projectile exec-path-from-shell pyenv-mode deft zenburn-theme which-key web-mode use-package try org-download org-bullets org-ac noflet jedi iedit hungry-delete htmlize flycheck expand-region elpy counsel cnfonts beacon ace-window))))

;; (custom-set-faces
;;  ;; custom-set-faces was added by Custom.
;;  ;; If you edit it by hand, you could mess it up, so be careful.
;;  ;; Your init file should contain only one such instance.
;;  ;; If there is more than one, they won't work right.
;;  '(aw-leading-char-face ((t (:inherit ace-jump-face-foreground :height 3.0)))))
(custom-set-faces
 ;; custom-set-faces was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(aw-leading-char-face ((t (:inherit ace-jump-face-foreground :height 3.0)))))
(put 'narrow-to-region 'disabled nil)


cnfonts/v4/profile1.el
;;; `cnfonts--custom-set-fontsnames' 列表有3个子列表，第1个为英文字体列表，第2个为中文字体列表，
;;; 第3个列表中的字体用于显示不常用汉字，每一个字体列表中，*第一个* *有效并可用* 的字体将被使用。
;;; 将光标移动到上述列表中，按 `C-c C-c' 可以测试字体显示效果。另外，用户可以通过命令
;;; `cnfonts-insert-fontname’ 来选择一个 *可用* 字体，然后在当前光标处插入其字体名称。
(setq cnfonts--custom-set-fontnames
      '(
        ("Monaco" "Hack" "Courier" "Courier New" "Consolas" "DejaVu Sans Mono" "Droid Sans Mono" "PragmataPro" "Ubuntu Mono" "Liberation Mono" "MonacoB" "MonacoB2" "MonacoBSemi" "Droid Sans Mono Pro" "Inconsolata" "Source Code Pro" "Lucida Console" "Envy Code R" "Andale Mono" "Lucida Sans Typewriter" "monoOne" "Lucida Typewriter" "Panic Sans" "Bitstream Vera Sans Mono" "HyperFont" "PT Mono" "Ti92Pluspc" "Excalibur Monospace" "Menlof" "Cousine" "Fira Mono" "Lekton" "M+ 1mn" "BPmono" "Free Mono" "Anonymous Pro" "ProFont" "ProFontWindows" "Latin Modern Mono" "Code 2002" "ProggyCleanTT" "ProggyTinyTT" "Iosevka Term" "Inconsolata-dz" "American Typewriter" "Menlo" "Anka/Coder Condensed" "Fantasque Sans Mono" "M+ 1m" "CamingoCode" "Office Code Pro" "Roboto Mono" "Input Mono" "Courier Prime Code" "NanumGothicCoding" "Monoid" "Edlo" "Iosevka" "Mononoki" "Robot Mono" "Fantasque" "Fira Code" "Go Mono" "Noto Sans Mono CJK" "InputMonoCompressed" "Hasklig" "Terminus" "FantasqueSansMono" "AnonymousPro" "3270" "Arimo" "D2Coding" "Inconsolata-g" "ProFont for Powerline" "Meslo" "Meslo Dotted" "Noto Mono" "Symbol Neu" "Tinos" "Space Mono" "SFMono Nerd Font")
        ("SimHei" "Hiragino Sans GB" "STFangsong" "华文楷体" "华文仿宋" "微软雅黑" "Noto Sans S Chinese Regular" "Microsoft Yahei" "Microsoft YaHei Mono" "Microsoft_Yahei" "Ubuntu Mono" "文泉驿等宽微米黑" "文泉驿等宽正黑" "黑体" "Source Han Sans SC" "Source Han Serif SC" "思源黑体 CN Regular" "思源黑体 CN Medium" "思源黑体 CN Normal" "思源宋体 CN" "思源宋体 CN Medium" "思源宋体 CN SemiBold" "文泉驿正黑" "文泉驿点阵正黑" "SimSun" "NSimSun" "FangSong" "KaiTi" "FangSong_GB2312" "KaiTi_GB2312" "LiSu" "YouYuan" "新宋体" "宋体" "楷体_GB2312" "仿宋_GB2312" "幼圆" "隶书" "STXihei" "STKaiti" "STSong" "STZhongsong" "FZShuTi" "FZYaoti" "STCaiyun" "STHupo" "STLiti" "STXingkai" "STXinwei" "方正姚体" "方正舒体" "方正粗圆_GBK" "华文中宋" "华文行楷" "华文彩云" "华文隶书" "华文细黑" "华文琥珀" "华文新魏" "微软简标宋" "微软简粗黑" "微软简老宋" "微软简魏碑" "微软简楷体" "微软简行楷" "微软简隶书" "微软简中圆" "微软简仿宋" "方正大黑_GBK" "方正大标宋_GBK" "方正报宋_GBK" "方正彩云_GBK" "方正综艺_GBK" "方正稚艺_GBK" "方正幼线_GBK" "方正准圆_GBK" "方正中等线_GBK" "方正中倩_GBK" "方正行楷_GBK" "方正细珊瑚_GBK" "方正细圆_GBK" "方正细倩_GBK" "方正新舒体_GBK" "方正新报宋_GBK" "方正小标宋_GBK" "方正姚体_GBK" "方正魏碑_GBK" "方正舒体__GBK" "方正细黑一_GBK" "方正细等线_GBK" "方正水柱_GBK" "方正宋黑_GBK" "方正宋三_GBK" "方正宋一_GBK" "方正隶二_GBK" "方正隶书_GBK" "方正胖娃_GBK" "方正美黑_GBK" "方正瘦金书_GBK" "方正平和_GBK" "方正少儿_GBK" "方正书宋_GBK" "方正黑体_GBK" "方正黄草_GBK" "方正隶变_GBK" "方正琥珀_GBK" "方正楷体_GBK" "方正康体_GBK" "方正华隶_GBK" "方正仿宋_GBK" "方正超粗黑_GBK" "方正粗宋_GBK" "方正粗倩_GBK" "方正徐静蕾字体" "方正黑体简体" "方正黄草简体" "方正隶二简体" "方正隶书简体" "方正铁筋隶书简体" "方正超粗黑简体" "方正行楷简体" "方正艺黑简体" "方正胖娃简体" "方正胖头鱼简体" "方正美黑简体" "方正综艺简体" "方正细黑一简体" "方正细等线简体" "方正细珊瑚简体" "方正细圆简体" "方正细倩简体" "方正粗活意简体" "方正粗宋简体" "方正粗圆简体" "方正粗倩简体" "方正稚艺简体" "方正祥隶简体" "方正硬笔行书简体" "方正硬笔楷书简体" "方正瘦金书简体" "方正琥珀简体" "方正流行体简体" "方正水黑简体" "方正水柱简体" "方正北魏楷书简体" "方正华隶简体" "方正卡通简体" "方正古隶简体" "方正启体简体" "方正大标宋简体" "方正大黑简体" "方正姚体简体" "方正宋一简体" "方正宋黑简体" "方正小标宋简体" "方正少儿简体" "方正平和简体" "方正幼线简体" "方正康体简体" "方正彩云简体" "方正报宋简体" "方正新报宋简体" "方正新舒体简体" "方正楷体简体" "方正毡笔黑简体" "方正剪纸简体" "方正准圆简体" "方正仿宋简体" "方正书宋简体" "方正中等线简体" "方正中倩简体" "方正黄草简" "方正魏碑简体" "方正隶变简体" "汉仪彩云体简" "汉仪彩蝶体简" "汉仪报宋简" "汉仪柏青体简" "汉仪白棋体简" "汉仪长宋简" "汉仪长美黑简" "汉仪长艺体简" "汉仪粗仿宋简" "汉仪粗圆简" "汉仪粗宋简" "汉仪粗黑简" "汉仪超粗黑简" "汉仪超粗宋简" "汉仪超粗圆简" "汉仪陈频破体简" "汉仪嘟嘟体简" "汉仪大宋简" "汉仪大隶书简" "汉仪大黑简" "汉仪方叠体简" "汉仪方隶简" "汉仪蝶语体简" "汉仪黛玉体简" "汉仪仿宋简" "汉仪哈哈体简" "汉仪橄榄体简" "汉仪海韵体简" "汉仪琥珀体简" "汉仪花蝶体简" "汉仪黑咪体简" "汉仪黑棋体简" "汉仪凌波体简" "汉仪家书简" "汉仪楷体简" "汉仪漫步体简" "汉仪火柴体简" "汉仪立黑简" "汉仪菱心体简" "汉仪萝卜体简" "汉仪书宋一简" "汉仪书宋二简" "汉仪书魂体简" "汉仪南宫体简" "汉仪咪咪体简" "汉仪清韵体简" "汉仪瘦金书简" "汉仪神工体简" "汉仪双线体简" "汉仪太极体简" "汉仪娃娃篆简" "汉仪水波体简" "汉仪水滴体简" "汉仪特细等线简" "汉仪舒同体简" "汉仪魏碑简" "汉仪小隶书简" "汉仪秀英体简" "汉仪细中圆简" "汉仪细圆简" "汉仪细等线简" "汉仪细行楷简" "汉仪行楷简" "汉仪醒示体简" "汉仪丫丫体简" "汉仪中楷简" "汉仪中等线简" "汉仪中黑简" "汉仪圆叠体简" "汉仪雁翎体简" "汉仪雪君体简" "汉仪雪峰体简" "汉仪中圆简" "汉仪中宋简" "汉仪中隶书简" "汉仪竹节体简" "汉仪字典宋简" "经典行楷简" "经典魏碑简" "经典空叠圆简" "经典仿宋简" "经典细标宋简" "经典细等线简" "经典中宋简" "经典标宋简" "经典粗仿黑" "经典粗黑简" "经典超圆简" "经典长宋简" "经典叠圆体简" "经典等线简" "经典粗圆简" "经典粗宋简" "经典楷体简" "经典空叠黑" "经典黑体简" "经典平黑简" "经典空趣体简" "经典美黑简" "经典隶变简" "经典图案字" "经典宋体简" "经典特宋简" "经典特黑简" "经典舒同体简" "经典趣体简" "经典圆叠黑" "经典细空艺" "经典细空黑" "经典行书简" "经典细宋简" "经典细隶书简" "经典中圆简" "经典圆体简" "经典综艺体简" "经典细圆简" "长城中圆体" "长城黑宋体" "长城黑体" "长城细圆体" "长城长宋体" "长城行楷体" "长城行书体" "长城美黑体" "长城细仿宋体" "长城粗魏碑体" "长城粗隶书体" "长城粗行楷体" "长城粗圆体" "长城特粗黑" "长城特粗宋体" "长城特粗圆体" "长城特圆体" "长城楷体" "长城新魏碑体" "长城新艺体" "长城报宋体" "长城小标宋体" "长城小姚体" "长城宋体" "长城大黑体" "长城大标宋体" "长城仿宋体" "长城中隶体" "长城中行书体")
        ("HanaMinB" "SimSun-ExtB" "MingLiU-ExtB" "PMingLiU-ExtB" "MingLiU_HKSCS-ExtB" "Hanazono Mincho" "Hanazono Mincho A" "Hanazono Mincho B" "Hanazono Mincho C" "Hanazono Mincho Ex" "Hanazono Mincho Ex A1" "Hanazono Mincho Ex A2" "Hanazono Mincho Ex B" "Hanazono Mincho Ex C" "Hanazono Mincho I")
        ))

;;; `cnfonts--custom-set-fontsizes' 中，所有元素的结构都类似：(英文字号 中文字号 EXT-B字体字号)
;;; 将光标移动到各个数字上，按 C-c C-c 查看光标处字号的对齐效果。
;;; 按 C-<up> 增大光标处字号，按 C-<down> 减小光标处字号。
(setq cnfonts--custom-set-fontsizes
      '(
        (9    10.5 10.0)
        (10   12.0 12.0)
        (11.5 14.0 13.5)
        (12.5 14.5 15.0)
        (14   16.5 16.5)
        (15   18.0 18.0)
        (16   20.0 19.5)
        (18   22.0 21.0)
        (20   24.0 24.0)
        (22   25.5 25.5)
        (24   28.5 28.5)
        (26   31.5 31.5)
        (28   33.0 33.0)
        (30   36.0 36.0)
        (32   39.0 39.0)
        ))


elpa/gnu-elpa-keyring-update-2019.3/gnu-elpa-keyring-update-autoloads.el
;;; gnu-elpa-keyring-update-autoloads.el --- automatically extracted autoloads
;;
;;; Code:
(add-to-list 'load-path (directory-file-name (or (file-name-directory #$) (car load-path))))

;;;### (autoloads nil "gnu-elpa-keyring-update" "gnu-elpa-keyring-update.el"
;;;;;;  (24269 51341 519749 400000))
;;; Generated autoloads from gnu-elpa-keyring-update.el

(defvar gnu-elpa-keyring-update--keyring (let ((kr (expand-file-name "etc/gnu-elpa.gpg-keyring" (file-name-directory load-file-name)))) (if (and load-file-name (file-readable-p kr)) kr "etc/gnu-elpa.gpg-keyring")))

(autoload 'gnu-elpa-keyring-update "gnu-elpa-keyring-update" "\
Import new GNU ELPA keys (if any) into package.el's keyring.

\(fn)" nil nil)
 (eval-after-load 'package
  `(and (bound-and-true-p package-user-dir)
        (file-directory-p package-user-dir)
        (let ((ts (expand-file-name
                   "gnu-elpa.timestamp"
                   (or (bound-and-true-p package-gnupghome-dir)
                       (expand-file-name "gnupg"
                                         package-user-dir))))
              (kr gnu-elpa-keyring-update--keyring))
          (and (file-writable-p ts)
               (file-readable-p kr)
               (file-newer-than-file-p kr ts)
               (gnu-elpa-keyring-update)))))

;;;***

;;;### (autoloads nil nil ("gnu-elpa-keyring-update-pkg.el") (24269
;;;;;;  51341 403060 700000))

;;;***

;; Local Variables:
;; version-control: never
;; no-byte-compile: t
;; no-update-autoloads: t
;; End:
;;; gnu-elpa-keyring-update-autoloads.el ends here


myinit.el
;;hide startup massage
(setq inhibit-startup-message t)
;; close
(setq make-backup-files nil)
;;hide tool bar
(tool-bar-mode -1)
;; use y-n-p over yes-no-p
(fset 'yes-or-no-p 'y-or-n-p)
;;reload buffer use F5
(global-auto-revert-mode 1)
(setq auto-revert-verbose nil)
(global-set-key (kbd "<f5>") 'revert-buffer)
;;list buffers
(defalias 'list-buffers 'ibuffer)
;;set command-key to meta
(if (eq system-type 'darwin)
    (progn (setq mac-option-modifier 'super)
           (setq mac-command-modifier 'meta)))
;; turn off Simultaneous Editing
(setq create-lockfiles nil)

;; all backups goto ~/.backups instead in the current directory
(setq backup-by-copying t
      backup-directory-alist
      '(("." . "~/.emacs.d/backup"))
      delete-old-versions t
      kept-new-versions 3
      kept-old-versions 1
      version-control t)

;;auto-save
(setq-default auto-save-timeout 15) ; 15秒无动作,自动保存
(setq-default auto-save-interval 100) ; 100个字符间隔, 自动保存

;;设置窗口位置为屏库左上角(0,0)
(set-frame-position (selected-frame) 0 0)
;;设置宽和高
(set-frame-width (selected-frame) 170)
(set-frame-height (selected-frame) 80)

;; 最大化窗口
;;(add-to-list 'default-frame-alist '(fullscreen . fullboth))
;;(add-hook 'window-setup-hook 'toggle-frame-maximized t)
;;(add-hook 'window-setup-hook 'toggle-frame-fullscreen t)

;;括号补全
(electric-pair-mode t)
;;自动换行
(add-hook 'org-mode-hook (lambda () (setq truncate-lines nil)))
;;开启行号
(global-linum-mode t)

;;(setq fonts
;;      (cond ((eq system-type 'darwin)     '("Monaco"    "STHeiti"))
;;            ((eq system-type 'gnu/linux)  '("Menlo"     "STHeiti"))
;;            ((eq system-type 'windows-nt) '("Consolas"  "Microsoft Yahei"))))
;;(set-face-attribute 'default nil :font
;;                    (format "%s:pixelsize=%d" (car fonts) 14))
;;(dolist (charset '(kana han symbol cjk-misc bopomofo))
;;  (set-fontset-font (frame-parameter nil 'font) charset
;;                    (font-spec :family (car (cdr fonts)))))
;;Fix chinese font width and rescale
;;(setq face-font-rescale-alist '(("Microsoft Yahei" . 1.2) ("WenQuanYi Micro HeiMono" . 1.2) ("STHeiti". 1.2)))

(global-hl-line-mode t)

(use-package beacon
  :ensure t
  :config
  (beacon-mode 1)
  ;;(setq beacon-color "#666600")
  )

(setq save-interprogram-paste-before-kill t)

(use-package hungry-delete
  :ensure t
  :config
  (global-hungry-delete-mode))

(use-package expand-region
  :ensure t
  :config
  (global-set-key (kbd "C-=") 'er/expand-region))

(use-package iedit
  :ensure t)

(use-package org-bullets
  :ensure t
  :config
  (add-hook 'org-mode-hook #'org-bullets-mode))

;; set org-file-apps?
;; (setq org-file-apps
;;       (append '(
;;                 ("\\.pdf\\'" . "evince %s")
;;                 ("\\.x?html?\\'" . "/usr/bin/firefox %s")
;;                 ) org-file-apps ))

;;key bind and some stuff
(global-set-key "\C-ca" 'org-agenda)
(setq org-agenda-start-on-weekday nil)
;;load agenda files
(setq org-agenda-files (list "~/Dropbox/org/*.org"))

;;use org-ac auto-complete for org
(use-package org-ac
  :ensure t
  :init (progn
          (require 'org-ac)
          (org-ac/config-default)
          ))
    ;;set key bind to C-c c
(global-set-key (kbd "C-c c") 'org-capture)

;;set org capture templates
(setq org-capture-templates
      '(("l" "Link" entry (file+headline "~/Dropbox/org/inbox.org" "Links")
         "* %? %^L %^g \n%T" :prepend t)
        ("t" "To-Do Item" entry (file+headline "~/Dropbox/org/inbox.org" "To-Do")
         "* TODO %?\n%u" :prepend t)
        ("n" "Note" entry (file+headline "~/Dropbox/org/inbox.org" "Notes")
         "* %u %? " :prepend t)
        ))

;;images support
(use-package org-download
  :ensure t
  :config
  (add-hook 'dired-mode-hook 'org-download-enable)
  (setq org-download-heading-lvl nil)
  (setq-default org-download-image-dir "~/Dropbox/org/images")
  )
;; set defauly img width in org
(setq org-image-actual-width 450)


;; (defun after-org-mode-loaded()
;;   (setq org-download-heading-lvl nil)
;;   (if buffer-file-name
;;       (let ((org-img-base-folder "~/Dropbox/org/images")
;;             (img-folder (format "%s/%s" org-img-base-folder (file-name-base buffer-file-name))))
;;         (setq-local org-download-image-dir img-folder))
;;     ))

;; use htmlize
(use-package htmlize
  :ensure t)
(require 'ox-publish)
(require 'ox-html)
(require 'ox-md)

;; (setq org-publish-project-alist
;;       '(
;;         ("blog-notes"
;;          :base-directory "~/Dropbox/org/notes/"
;;          :base-extension "org"
;;          :publishing-directory "~/Dropbox/org/html/"
;;          :recursive t
;;          :publishing-function org-html-publish-to-html
;;          :headline-levels 2
;;          :auto-preamble t
;;          :section-numbers nil
;;          :author "Enke"
;;          :email "enke000@gmail.com"
;;          :auto-sitemap t                  ; 自动生成 sitemap.org 文件
;;          :sitemap-filename "sitemap.org"  ; ... call it sitemap.org (it's the default)...
;;          :sitemap-title "Sitemap"         ; ... with title 'Sitemap'.
;;          :sitemap-sort-files anti-chronologically
;;          :sitemap-file-entry-format "%d %t"
;;          )
;;         ("blog-static"
;;          :base-directory "~/Dropbox/org/notes/"
;;          :base-extension "css|js|png|jpg|gif|pdf|mp3|ogg|swf"
;;          :publishing-directory "~/Dropbox/org/html/"
;;          :recursive t
;;          :publishing-function org-publish-attachment
;;          )
;;         ("blog" :components ("blog-notes" "blog-static"))
;;         ))

;;deft emacs
(use-package deft
  :ensure t
  :bind ("C-c d" . deft)
  :commands (deft)
  :init (setq deft-directory "~/Dropbox/org/notes"
              deft-extensions '("org")
              deft-text-mode 'org-mode
              deft-recursive t
              deft-file-naming-rules '((noslash . "-"))
              deft-text-mode 'org-mode
              deft-use-filter-string-for-filename t
              deft-strip-summary-regexp
              (concat "\\("
                      "[\n\t]" ;; blank
                      "\\|^#\\+[[:upper:]_]+:.*$" ;; org-mode metadata
                      "\\|^#\\+[[:alnum:]_]+:.*$" ;; org-mode metadata
                      "\\)")))

(use-package try
  :ensure t)

(use-package which-key
  :ensure t
  :config
  (which-key-mode))

(setq indo-enable-flex-matching t)
(setq ido-everywhere t)
(ido-mode 1)

(use-package ace-window
  :ensure t
  :init
  (progn
    (global-set-key [remap other-window] 'ace-window)
    ;;change faces of ace-window
    (custom-set-faces
     '(aw-leading-char-face
       ((t (:inherit ace-jump-face-foreground :height 3.0)))))
    ))

(use-package counsel
  :ensure t
  :bind
  (("M-y" . counsel-yank-pop)
   :map ivy-minibuffer-map
   ("M-y" . ivy-next-line)))

(use-package swiper
  :ensure try
  :bind(("C-s" . swiper)
	("C-r" . swiper)
	("C-c C-r" . ivy-resume)
	("M-x" . counsel-M-x)
	("C-x C-f" . counsel-find-file))
  :config
  (progn
    (ivy-mode 1)
    (setq ivy-use-virtual-buffers t)
    (setq ivy-display-style 'fancy)
    (define-key read-expression-map (kbd "C-r") 'counsel-expression-history)
    ))

(use-package avy
  :ensure t
  :bind("M-s" . avy-goto-char))

(use-package auto-complete
  :ensure t
  :init
  (progn
    (ac-config-default)
    (global-auto-complete-mode t)
    ))

(use-package zenburn-theme
  :ensure t
  :config (load-theme 'zenburn t))

;;(use-package htmlize
  ;;:ensure t)

(use-package flycheck
  :ensure t
  :init
  (global-flycheck-mode t))

(use-package elpy
  :ensure t
  :config
      (elpy-enable))

;;(setq python-shell-interpreter "~/.pyenv/shims/python")

(use-package pyenv-mode
  :ensure t)

(use-package jedi
  :ensure t
  :init
  (add-hook 'python-mode-hook 'jedi:setup)
      (add-hook 'python-mode-hook 'jedi:ac-setup))

(use-package exec-path-from-shell
  :ensure t
  :init
  (when (memq window-system '(mac ns x))
    (exec-path-from-shell-initialize)))

(use-package yasnippet
  :ensure t
  :init
  (yas-global-mode 1))

(use-package web-mode
  :ensure t
  :config
  (add-to-list 'auto-mode-alist '("\\.html?\\'" . web-mode))
  (setq web-mode-engines-alist
	'(("django"    . "\\.html\\'")))
  (setq web-mode-ac-sources-alist
	'(("css" . (ac-source-css-property))
	  ("html" . (ac-source-words-in-buffer ac-source-abbrev))))

  (setq web-mode-enable-auto-closing t)
  (setq web-mode-enable-auto-quoting t)) ; this fixes the quote problem I mentioned

(defun now ()
  "Insert string for the current time formatted like '2:34 PM' or 1507121460"
  (interactive)                 ; permit invocation in minibuffer
  ;;(insert (format-time-string "%D %-I:%M %p")))
  (insert (format-time-string "%-I:%M %p")))
  ;;(insert (format-time-string "%02y%02m%02d%02H%02M%02S")))
  ;;(insert (format-time-string "%02y%02m%02d%02H%02M")))

(global-set-key "\C-xt" 'now)

(global-set-key (kbd "C-x C-b") 'ibuffer)
(setq ibuffer-saved-filter-groups
      (quote (("default"
               ("dired" (mode . dired-mode))
               ("org" (name . "^.*org$"))

               ("web" (or (mode . web-mode) (mode . js2-mode)))
               ("shell" (or (mode . eshell-mode) (mode . shell-mode)))
               ("mu4e" (name . "\*mu4e\*"))
               ("programming" (or
                               (mode . python-mode)
                               (mode . c++-mode)))
               ("emacs" (or
                         (name . "^\\*scratch\\*$")
                         (name . "^\\*Messages\\*$")))
               ))))
(add-hook 'ibuffer-mode-hook
          (lambda ()
            (ibuffer-auto-mode 1)
            (ibuffer-switch-to-saved-filter-groups "default")))

;; don't show these
;;(add-to-list 'ibuffer-never-show-predicates "zowie")
;; Don't show filter groups if there are no buffers in that group
(setq ibuffer-show-empty-filter-groups nil)

;; Don't ask for confirmation to delete marked buffers
(setq ibuffer-expert t)

(use-package emmet-mode
  :ensure t
  :config
  (add-hook 'sgml-mode-hook 'emmet-mode) ;; Auto-start on any markup modes
  (add-hook 'web-mode-hook 'emmet-mode) ;; Auto-start on any markup modes
  (add-hook 'css-mode-hook  'emmet-mode) ;; enable Emmet's css abbreviation.
  )
